pkg/services/provisioning/provisioning.go

```diff
@@ -13,35 +14,119 @@ import (
 )
 
 func init() {
-	registry.RegisterService(&ProvisioningService{})
+	registry.RegisterService(NewProvisioningServiceImpl(
+		func(path string) (dashboards.DashboardProvisioner, error) {
+			return dashboards.NewDashboardProvisionerImpl(path)
+		},
+		notifiers.Provision,
+		datasources.Provision,
+	))
 }
 
-type ProvisioningService struct {
-	Cfg *setting.Cfg `inject:""`
+type ProvisioningService interface {
+	ProvisionDatasources() error
+	ProvisionNotifications() error
+	ProvisionDashboards() error
 }
 
-func (ps *ProvisioningService) Init() error {
-	datasourcePath := path.Join(ps.Cfg.ProvisioningPath, "datasources")
-	if err := datasources.Provision(datasourcePath); err != nil {
-		return fmt.Errorf("Datasource provisioning error: %v", err)
+func NewProvisioningServiceImpl(
+	newDashboardProvisioner func(string) (dashboards.DashboardProvisioner, error),
```

Comment:
Any specific reason for using func's here? Can't seem to see the point of this. Oh they're used as factory methods, right? Maybe a good idea to extract the func's to types then, like `DashboardProvisionerFactory` or something.

Reply:
Yes exactly. There is a bit of setup in the `NewDashboardProvisionerImpl` that needs to be done so I needed the factory function. But will do the factory function type. The typing was tedious but it did not occur to me I could do that.